[CODE CAMP] AbstractBeam

IvarsKarpics opened the pull request:
Question to all:
Should we define some rules how the values are returned. For example HWR.beamline.beam.get_size() should return a list with two floats or two floats?
Now I changed the code to return two floats.

rhfogh commented:
I would vote for standardising on two floats - which is actually a *tuple* of two floats. There is generally not that much point to passing around modifiable data structures like lists unless you need them for some reason. But eitehr woudl work, of course.

rhfogh commented on HardwareObjects/QtGraphicsManager.py:
```diff
@@ -275,8 +275,9 @@ def init(self):
-            self.beam_info_dict = HWR.beamline.beam.get_beam_info()
-            self.beam_position = HWR.beamline.beam.get_beam_position()
+            self.beam_info_dict = HWR.beamline.beam.get_info_dict()
+            pos_x, pos_y = HWR.beamline.beam.get_screen_position()
```
This is fine. But, FYI,
    self.beam_position = HWR.beamline.beam.get_screen_position()
 would also work .

rhfogh commented on HardwareObjects/XMLRPCServer.py:
```diff
@@ -523,22 +523,22 @@ def flux(self):
+        HWR.beamline.beam.aperture.moveToPosition(pos_name)
         t0 = time.time()
-        while HWR.beamline.beam.aperture_hwobj.getState() == "MOVING":
+        while HWR.beamline.beam.aperture.getState() == "MOVING":
```
Could this not be done with aperture.wait_ready()?

IvarsKarpics replied:
hmm, I would suggest to open a new issue for AbstractAperture

rhfogh replied:
Not sure it needs an issue. The wait_ready function is already there - it is inherited from BaseHardwareObjects.HardwareObjectMixin. That makes it just a question of using the existing function as opposed to replicating the behaviour elsewhere.

IvarsKarpics replied:
Yes, I know, but actual wait_ready is not implemented.

rhfogh replied:
I think it is, actually. The code in HardwareObjectMixin is not abstract, and should be functional. It only requires that self._ready_event() is set when the object becomes ready. The function code is:

    def wait_ready(self, timeout=None):
        """Wait till object is ready.

        If timeout == 0: return at once and do not wait;
        if timeout is None: wait forever."""
        if timeout != 0:
            with gevent.Timeout(
                timeout, RuntimeError("Timeout waiting for status ready")
            ):
                self._ready_event.wait(timeout=timeout)

rhfogh commented on HardwareObjects/abstract/AbstractBeam.py:
```diff
@@ -0,0 +1,104 @@
+
+        :return: tuple
+        """
+        return self._divergence[0], self._divergence[1]
```
return tuple(self._divergence) woudl also work.

Given that self._divergence is a tuple, even the follwoing would be perfectly safe:

return self._divergence

IvarsKarpics replied:
ok

rhfogh commented on HardwareObjects/abstract/AbstractBeam.py:
```diff
@@ -0,0 +1,104 @@
+    def aperture(self):
+        return self._aperture
+
+    @property
```
DOsukcntatino giving the data type woudl be nice, especially for a plural parameter name

beteva replied:
This is precisely an issue we will address during the ''documentation'' developer's meeting. Meanwhile if you use pylint, it complains that the documentation is missing ;-)

rhfogh commented on HardwareObjects/abstract/AbstractBeam.py:
```diff
@@ -0,0 +1,104 @@
+        """
+        return self._shape
+
+    def get_info_dict(self):
```
Could we do 
return self.info_dict.copy()
It si always bad t6o export mutable internal atributes.

rhfogh commented on HardwareObjects/abstract/AbstractFlux.py:
```diff
@@ -51,12 +50,18 @@ class AbstractFlux(HardwareObject):
+        """Get flux at current transmission in units of photons/s"""
+        return self._value
 
     def get_flux(self):
```
Presubmbly temporary, pending removel (?)

rhfogh commented on HardwareObjects/abstract/AbstractFlux.py:
```diff
@@ -51,12 +50,18 @@ class AbstractFlux(HardwareObject):
     def get_flux(self):
-        """Get flux at current transmission in units of photons/s"""
-        raise NotImplementedError
+        return self._value
```
Presumably temoporarily, pending removel (?)

rhfogh commented on HardwareObjects/abstract/AbstractFlux.py:
```diff
@@ -51,12 +50,18 @@ class AbstractFlux(HardwareObject):
 
+        self._value = None
         self.dose_rate_per_photon_per_mmsq = dose_rate_per_photon_per_mmsq
+        self.measured_flux_list = []
```
Can't we store the internal value as as tuple, and convert to tuple when setting? It is a long-term risk to import and export mutable data structures this way.

IvarsKarpics replied:
maybe I will revert AbstractFlux to the master. It do not belong to the PR

rhfogh commented on HardwareObjects/abstract/AbstractFlux.py:
```diff
@@ -71,17 +76,28 @@ def get_dose_rate(self, energy=None):
+    def get_measured_flux_list(self):
+        return self.measured_flux_list
+
+    def set_measured_flux_list(self, measured_flux_list):
```
Convert to tuple before setting

rhfogh commented on HardwareObjects/abstract/AbstractFlux.py:
```diff
@@ -51,12 +50,18 @@ class AbstractFlux(HardwareObject):
+        self._value = None
         self.dose_rate_per_photon_per_mmsq = dose_rate_per_photon_per_mmsq
+        self.measured_flux_list = []
+        self.measured_flux_dict = None
```
If intgernal, they shoudl maybe start with an underscore. If exported, what is the context, and how do we avoid that they are modified from the outside?

IvarsKarpics replied:
Could we leave AbstractFlux as it is? I just added to the commit because it contained changes related to the beam. I am ok to discuss this in a separate issue/pr

rhfogh replied:
OK. Again, this would be future refactoring. We do not need to do all the refactoring at once.

rhfogh commented on HardwareObjects/abstract/AbstractXRFSpectrum.py:
```diff
@@ -63,15 +63,27 @@ def start_spectrum(
                 if not os.path.exists(spectrum_directory):
                     os.makedirs(spectrum_directory)
+            except OSError as diag:
+                logging.getLogger().error(
```
I believe that 
logging.getLogger.exception() 
will add the print of the latest error withiout you having to do it manually.

beteva commented:
agree with @rhfogh, two floats is actually a tuple and I vote for returning a tuple when posiible.

rhfogh commented on configuration/embl_hh_p14/eh1/slits.xml:
```diff
@@ -0,0 +1,44 @@
+<object class="EMBLSlitBox">
```
Future refactoring (so NOT for changing now):
Could we call this 'slitbox' instead of 'slits'? For [plural attribute names I always think that we ahve a colelction of slits, ratehr than a single object.

IvarsKarpics commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
- Added 28
           

Complexity increasing per file
==============================
- HardwareObjects/mockup/BeamMockup.py  3
- HardwareObjects/abstract/AbstractBeam.py  1
         

Complexity decreasing per file
==============================
+ HardwareObjects/abstract/AbstractXRFSpectrum.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/mxcube/HardwareRepository/pullRequest?prid=5127064&bid=16851680)

marcus-oscarsson commented:
I think we spoke briefly about this a few meetings ago and Antonia and I already added an AbstractBeam as part of our fixes that we've said that we wanted to make a PR of. So it would maybe be good if we had the opportunity to make that PR and then merge this after ?

I guess we should have made a comment on the issue to let you know, but I think we spoke about ... In anyways sorry for the inconvenience but we cant merge right away

IvarsKarpics commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
- Added 30
           

Complexity increasing per file
==============================
- HardwareObjects/abstract/AbstractBeam.py  1
- HardwareObjects/mockup/BeamMockup.py  3
         

Complexity decreasing per file
==============================
+ HardwareObjects/abstract/AbstractXRFSpectrum.py  -1
         

Clones added
============
- HardwareObjects/MD3UP.py  1
- HardwareObjects/ESRF/ESRFMultiCollect.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/mxcube/HardwareRepository/pullRequest?prid=5127064&bid=16851680)

IvarsKarpics commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
- Added 36
           

Complexity increasing per file
==============================
- HardwareObjects/mockup/BeamMockup.py  3
- HardwareObjects/abstract/AbstractBeam.py  1
         

Complexity decreasing per file
==============================
+ HardwareObjects/abstract/AbstractXRFSpectrum.py  -1
         

Clones added
============
- HardwareObjects/MD3UP.py  1
- HardwareObjects/ESRF/ESRFMultiCollect.py  1
         

Clones removed
==============
+ HardwareObjects/ALBA/ALBABeamInfo.py  -3
+ HardwareObjects/abstract/AbstractCollect.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/mxcube/HardwareRepository/pullRequest?prid=5127064&bid=16851680)

IvarsKarpics commented:
Hi @marcus-oscarsson , @beteva , @rhfogh and others,

 I tried to merge the AbstractBeam and would like to hear your comments. I will change all subclasses after an overall agreement.

Cheers,
Ivars

IvarsKarpics commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
- Added 36
           

Complexity increasing per file
==============================
- HardwareObjects/abstract/AbstractBeam.py  1
- HardwareObjects/mockup/BeamMockup.py  3
         

Complexity decreasing per file
==============================
+ HardwareObjects/abstract/AbstractXRFSpectrum.py  -1
         

Clones added
============
- HardwareObjects/MD3UP.py  1
- HardwareObjects/ESRF/ESRFMultiCollect.py  1
         

Clones removed
==============
+ HardwareObjects/abstract/AbstractCollect.py  -1
+ HardwareObjects/ALBA/ALBABeamInfo.py  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/mxcube/HardwareRepository/pullRequest?prid=5127064&bid=16851680)

marcus-oscarsson commented:
@beteva is on vacation and will be back next week, so I think we need to wait for her last comments

marcus-oscarsson commented:
I noticed that there is sometimes beam.get_beam_position_on_screen and beam.get_screen_position whats the difference ?

IvarsKarpics commented:
Hi @marcus-oscarsson ,

As I said the PR is in progress and if we agree on the AbstractBeam then I will fix all other related classes.

marcus-oscarsson commented:
Oh, I see great !

rhfogh commented:
There is  one outstanding point: We should return tuples (not lists) for mulitple values, as @IvarsKarpics  proposes,  and this does not seem to be done overall yet
Apart form that I am fine with it.

What are we waiting for on this one?

marcus-oscarsson commented on HardwareObjects/Microdiff.py:
```diff
@@ -452,7 +452,7 @@ def move_to_beam(self, x, y):
         else:
             try:
-                beam_pos_x, beam_pos_y = HWR.beamline.beam.get_beam_position()
+                beam_pos_x, beam_pos_y = HWR.beamline.beam.get_screen_position()
```
@IvarsKarpics so I guess we will move this to SampleView when we get to that ?

IvarsKarpics replied:
maybe this change we can do after merging this PR in the next PR related to the sample view?

marcus-oscarsson replied:
Sure, just a comment not to forget to discuss it :)

marcus-oscarsson commented on HardwareObjects/XMLRPCServer.py:
```diff
@@ -525,19 +525,19 @@ def flux(self):
+        HWR.beamline.beam.aperture.moveToPosition(pos_name)
         t0 = time.time()
-        while HWR.beamline.beam.aperture_hwobj.get_state() == "MOVING":
+        while HWR.beamline.beam.aperture.get_state() == "MOVING":
```
This would break if aperture is an AbstractActuator, it would probably be better to check for HardwareObjectState.BUSY ?

marcus-oscarsson replied:
Sorry my mistake, replace AbstractActuator with simply HardwareObject :)

IvarsKarpics replied:
ok

marcus-oscarsson commented on HardwareObjects/XMLRPCServer.py:
```diff
@@ -525,19 +525,19 @@ def flux(self):
 
     def get_aperture_list(self):
-        return HWR.beamline.beam.aperture_hwobj.getPredefinedPositionsList()
+        return HWR.beamline.beam.aperture.getPredefinedPositionsList()
```
I know that it was not changed by you, but getPredefinedPositionsList should be camel case and I think its get_position_list or get_diameter_size_list now ?

IvarsKarpics replied:
ok, I will change

marcus-oscarsson commented on test/pytest/test_beam.py:
```diff
@@ -0,0 +1,42 @@
+#
```
There are no tests for the setters, it would probably be interesting

marcus-oscarsson commented:
This starts to look really nice, thanks !

I agree with @rhfogh that it would be nice if it is possible to use the set_value mechanism provided by AbstractActuator or inherit HardwareObject instead.

How is the set_value thought to be used by the way, there is no beam defining device being used in the end. It is not clear to me which part to override to supply the logic that decides which device to use ?

I could imagine that a method that takes the size and the shape could return the definer to be used, this could either be done in configuration as a default solution for the simple cases or with a method that implements the decision logic ? Or did I misunderstood something ?

beteva commented:
For me it is more logical that AbstractBeam inherits from HardwareObject, as there is not much what could  be used from AbstarctActuator methods. I believe it is even misleading, if we consider it as an actuator. Anyhow, I do not mind what it will be, but  I agree with @rhfogh that if it inherits from AbstractActuator, the method signatures should be the same.
I also agree with @marcus-oscarsson that the mechanism of a definer is unclear. I would also like to understand what the properties aperture, slits and definer are - strings, objects ?

IvarsKarpics commented:
Hi @rhfogh , @beteva  and @marcus-oscarsson 

I incorporated your suggestions:
* AbstractBeam inherits HardwareObject, removed set_value and get_value. There is no single device, equipment that can set beam properties. Size and shape is evaluated based on the slits gaps and aperture diameter.
* Fixed some camel cases to snake cases. However XMLRPCServer I would leave for a separate PR.
* Added nice setter test. a) set slits to 1mm and loop through all apertures and see if the beam size is equal to aperture size. b) selects max aperture diameter and sets smaller slits gaps and checks if target size is equal to current size.

TODO:
* move get_beam_position_on_screen to AbstractSampleView

rhfogh commented on HardwareObjects/abstract/AbstractBeam.py:
```diff
@@ -23,12 +23,13 @@
 """
 
-__copyright__ = """ Copyright © 2019 by the MXCuBE collaboration """
+__copyright__ = """MXCuBE collaboration"""
```
I thought that the agreed string was  
"Copyright © 2019-20120 by the MXCuBE collaboration "
we might use different starting years (even if this file was created 2019), but do we not want a range of years?

The copyright symbol should be harmless as long as you have the encoding set in the file header. Not sure if we have a cnsensus here, though.

beteva replied:
I thought it was the case. Whnecer i need to do a change in a file, I also set the range of dates to be  2010-2020.

rhfogh commented on HardwareObjects/abstract/AbstractBeam.py:
```diff
@@ -93,37 +123,44 @@ def get_available_size(self):
+        """
+        return self._beam_info_dict.copy()
+
     def update_value(self, value=None):
```
Should this not be removed, now there is no 'value'? 
We may need update_values(lural, or reemit, or whatever), but that is another story.

IvarsKarpics replied:
Well, we still need a single method that returns all beam parameters.

rhfogh commented:
@All
It looks to me like there might be a use for more setting and/or updating in AbstractBeam. But maybe that is intended for a later PR?

I can find no setters, in AbstractBeam. I guess that you may be supposed to set individual parameters directly on the apertures, slits, or beamdefiner, but it woud be nice if there was some specificatoin of a beamline-independent way of setting beam size or shape.
Also, might it be possible to show the intended implementation for getting the values (at least in simple cases) in AbstractBeam, something on the lines of putting an evaluate_beam_info into AbstractBeam (possible as an abstract method)? 
Or is this going to be so beamline-specific that even the implementation will vary from place to place?

This (unlike specicfic comments) should not hold up merging, but I thought it was worth raising for discussion.

rhfogh commented on HardwareObjects/mockup/BeamMockup.py:
```diff
@@ -0,0 +1,132 @@
+        self._beam_info_dict["size_y"] = size_y
+        self._beam_info_dict["shape"] = self._beam_shape
+
+    def emit_beam_info_change(self):
```
Might it be better to check the beam width and height against hte previous value rathet than the mockup default (even in a mockup people might test haveing successive changes)?

Also would this function be a candidate for implementing fully in AbstractBeam?

IvarsKarpics replied:
Hi @rhfogh and others. Following your comments I will improve the AbstracBeam by:
* moving `evaluate_beam_info` and `emi_beam_info_changed` to AbstractBeam,
* add a setter method like `set_beam_params(width=None, height=None, shape=Enum)`

marcus-oscarsson commented:
Thanks again for this nice work, this is a quite tricky one considering that it involves a few different types of devices that all change the beam in some way.

I agree with @rhfogh:

> but it woud be nice if there was some specificatoin of a beamline-independent way of setting beam size or shape

Basically a get/set beam_size and shape would be nice, either as two separate functions or as one function (in my opinion the later). There could be a simple default implementation but the more complex cases would need to be handled with a specific implementation for each beamline.

marcus-oscarsson commented on HardwareObjects/abstract/AbstractBeam.py:
```diff
@@ -48,38 +49,67 @@ class AbstractBeam(HardwareObject):
+
+        self._aperture = None
+        self._slits = None
+        self._definer = None
```
What is definer physically ?

rhfogh replied:
@IvarsKarpics 
Sounds good.

IvarsKarpics replied:
definer name comes from @beteva . It could be focusing optics, CRLs or any other device, equipment that changes beam. For now it is empty and may be filled later.

beteva commented:
Looks OK for me, except for the tests. Thanks @IvarsKarpics

IvarsKarpics commented:
Hi @beteva  and others. The beam test is passing and other ones are failing:
```
test/pytest/test_beam.py .....                                           [  6%]
test/pytest/test_dataobject.py ......                                    [ 13%]
test/pytest/test_detector.py .....                                       [ 19%]
test/pytest/test_detector_distance.py .............                      [ 35%]
test/pytest/test_energy.py .....FF......                                 [ 51%]
test/pytest/test_procedure.py ...                                        [ 55%]
test/pytest/test_sample_changer.py FFF......                             [ 66%]
test/pytest/test_shutter.py ......FF......                               [ 83%]
test/pytest/test_transmission.py .....FF......                           [100%]
```
One can fix them in next PRs

marcus-oscarsson commented:
Looks good to me, lets have a look a the tests in another PR and make the work as we add the final touches of the wother PR's.

Thanks alot  !

IvarsKarpics commented:
Merci
